File: lang/java/HTMLTable.java
// See Gaddis, Starting out with C++, Early Objects, Ninth Edition, pp. 800-804

import java.io.*;
import java.util.*;

public class HTMLTable
{
    private Vector<String> headers;
    private Vector<Vector<String>> rows;
    
    HTMLTable()
    {
        headers = new Vector<String>();
        rows = new Vector<Vector<String>>();
    }
    
    void setHeaders(String[] headers)
    {
        this.headers.addAll(Arrays.asList(headers));
    }
    
    void addRow(String[] row)
    {
        rows.add(new Vector<String>(Arrays.asList(row)));
    }
    
    void writeRow(PrintStream out, String tag, Vector<String> row)
    {
        out.println("<tr>");
        for (int k = 0; k < headers.size(); k++) {
            out.print("<" + tag + "> " + row.get(k) + " </" + tag + "> ");
        }
        out.println("\n</tr>");
    }

    static void putTo(PrintStream out, HTMLTable htmlTable)
    {
        out.println("<table border = \"1\">");
        htmlTable.writeRow(out, "th", htmlTable.headers);
        for (int r = 0; r < htmlTable.rows.size(); r++) {
            htmlTable.writeRow(out, "td", htmlTable.rows.get(r));
        }
        out.println("</table>");
    }
    
    public static void main(String[] args)
    {

        String[] headers = { "Name", "Address", "Phone" };
        String[] person1 = { "Mike Sans", "1215 Mills St", "630-728-1293" };
        String[] person2 = { "Natasha Upenski", "513 Briarcliff Ln", "412-672-1004" };
        
        HTMLTable hTable = new HTMLTable();
        hTable.setHeaders(headers);
        hTable.addRow(person1);
        hTable.addRow(person2);

        putTo(System.out, hTable);
    }

}

File: kotlin/HTMLTable.kt
// See Gaddis, Starting out with C++, Early Objects, Ninth Edition, pp. 800-804

import java.io.*
import java.util.*

class HTMLTable internal constructor() {
    private val headers: Vector<String>
    private val rows: Vector<Vector<String>>

    init {
        headers = Vector()
        rows = Vector()
    }

    internal fun setHeaders(headers: Array<String>) {
        this.headers.addAll(Arrays.asList(*headers))
    }

    internal fun addRow(row: Array<String>) {
        rows.add(Vector(Arrays.asList(*row)))
    }

    internal fun writeRow(out: PrintStream, tag: String, row: Vector<String>) {
        out.println("<tr>")
        for (k in headers.indices) {
            out.print("<$tag>${row[k]}</$tag>")
        }
        out.println("\n</tr>")
    }

    companion object {

        internal fun putTo(out: PrintStream, htmlTable: HTMLTable) {
            out.println("<table border = \"1\">")
            htmlTable.writeRow(out, "th", htmlTable.headers)
            for (r in htmlTable.rows.indices) {
                htmlTable.writeRow(out, "td", htmlTable.rows[r])
            }
            out.println("</table>")
        }

        @JvmStatic
        fun main(args: Array<String>) {

            val headers = arrayOf("Name", "Address", "Phone")
            val person1 = arrayOf("Mike Sans", "1215 Mills St", "630-728-1293")
            val person2 = arrayOf("Natasha Upenski", "513 Briarcliff Ln", "412-672-1004")

            val hTable = HTMLTable()
            hTable.setHeaders(headers)
            hTable.addRow(person1)
            hTable.addRow(person2)

            putTo(System.out, hTable)
        }
    }

}